moquette/broker/src/main/java/io/moquette/broker/SessionRegistry.java

```diff
@@ -145,82 +139,58 @@ private void recreateSessionPool() {
         final SessionCreationResult creationResult;
         if (oldSession.disconnected()) {
-            if (newIsClean) {
-                boolean result = oldSession.assignState(SessionStatus.DISCONNECTED, SessionStatus.CONNECTING);
-                if (!result) {
-                    throw new SessionCorruptedException("old session was already changed state");
-                }
-
-                // case 2
-                // publish new session
-                dropQueuesForClient(clientId);
-                unsubscribe(oldSession);
-                copySessionConfig(msg, oldSession);
-
-                LOG.trace("case 2, oldSession with same CId {} disconnected", clientId);
-                creationResult = new SessionCreationResult(oldSession, CreationModeEnum.CREATED_CLEAN_NEW, true);
-            } else {
-                final boolean connecting = oldSession.assignState(SessionStatus.DISCONNECTED, SessionStatus.CONNECTING);
-                if (!connecting) {
-                    throw new SessionCorruptedException("old session moved in connected state by other thread");
-                }
-                // case 3
-                reactivateSubscriptions(oldSession, username);
-
-                LOG.trace("case 3, oldSession with same CId {} disconnected", clientId);
-                creationResult = new SessionCreationResult(oldSession, CreationModeEnum.REOPEN_EXISTING, true);
+            final boolean connecting = oldSession.assignState(SessionStatus.DISCONNECTED, SessionStatus.CONNECTING);
+            if (!connecting) {
+                throw new SessionCorruptedException("old session moved in connected state by other thread");
             }
-        } else {
-            // case 4
-            LOG.trace("case 4, oldSession with same CId {} still connected, force to close", clientId);
-            oldSession.closeImmediately();
-            //remove(clientId);
-            creationResult = new SessionCreationResult(newSession, CreationModeEnum.DROP_EXISTING, true);
-        }
+            // case 2
+            reactivateSubscriptions(oldSession, username);
 
-        if (creationResult.mode == CreationModeEnum.DROP_EXISTING) {
-            LOG.debug("Drop session of already connected client with same id");
-            if (!pool.replace(clientId, oldSession, newSession)) {
-                //the other client was disconnecting and removed it's own session
-                pool.put(clientId, newSession);
-            }
+            LOG.trace("case 2, oldSession with same CId {} disconnected", clientId);
+            creationResult = new SessionCreationResult(oldSession, CreationModeEnum.REOPEN_EXISTING, true);
         } else {
-            LOG.debug("Replace session of client with same id");
-            if (!pool.replace(clientId, oldSession, oldSession)) {
-                throw new SessionCorruptedException("old session was already removed");
-            }
+            // case 3
```

Comment:
are these cases supposed to correspond to the old `// case 3`, `// case 4`, etc...?

I guess not since you've moved the `newIsClean` outside of this method

Reply:
They don't directly translate, I suppose. The old cases aren't clearly defined and IMO it's very difficult to reason about. Whereas the new cases are actually clearly defined.

Old cases:
1. Brand new session. Could be clean or persistent
2. Old session disconnected, new session is clean
3. Old session disconnected, new session is persistent
4. Old session connected, could be clean or persistent

New cases:
1. Session is clean or new
2. Session is persistent, old session is disconnected
3. Session is persistent, old session is connected

Case 1 now encompasses a little bit of old case 1, 2, and 4. New case 2 translates directly to old case 3. New case 3 encompasses half of old case 4.

Reply:
Let me rephrase. The old cases are defined, but it is difficult to correctly handle Session creation. There are only two things that matter when it comes to reconstructing a new session and I believe the new cases better align with this.

1. Is the session clean? If yes, throw away all of the old stuff and create a brand new, empty session. If the old one doesn't exist, you have no choice. Create a new one.
2. Session is persistent and the broker has an old Session. Resume the old session. If there is an old connection, disconnect it.